src/adapter/src/session.rs

```diff
@@ -103,6 +106,19 @@ pub struct Session<T = mz_repr::Timestamp> {
     transaction: TransactionStatus<T>,
     pcx: Option<PlanContext>,
     user: User,
+    /// The role ID of the current user.
+    ///
+    /// Invariant: role_id must be `Some` after the user has
+    /// successfully connected to and authenticated with Materialize.
+    ///
+    /// Prefer using this value over [`Self.user.name`].
+    //
+    // It would be better for this not to be an Option, but the
+    // `Session` is initialized before the user has connected to
+    // Materialize and is able to look up the `RoleId`. The `Session`
+    // is also used to return an error when no role exists and
+    // therefore there is no valid `RoleId`.
+    role_id: Option<RoleId>,
```

Comment:
I spent a lot of time trying to make this generic instead of using an option. Ultimately I couldn't get it to work, though I'm all ears if people have suggestions.

Reply:
This is the attempt if you're curious: https://github.com/MaterializeInc/materialize/compare/main...jkosh44:materialize:experimental-role-id